source/fuzz/transformation_equation_instruction.cpp

```diff
@@ -64,61 +65,93 @@ bool TransformationEquationInstruction::IsApplicable(
+  switch (opcode) {
+    case SpvOpConvertUToF:
+    case SpvOpConvertSToF: {
+      if (message_.in_operand_id_size() != 1) {
+        return false;
+      }
 
-  opt::Instruction::OperandList in_operands;
-  std::vector<uint32_t> rhs_id;
-  for (auto id : message_.in_operand_id()) {
-    in_operands.push_back({SPV_OPERAND_TYPE_ID, {id}});
-    rhs_id.push_back(id);
-  }
+      const auto* type = ir_context->get_type_mgr()->GetType(
+          fuzzerutil::GetTypeId(ir_context, message_.in_operand_id(0)));
+      if (!type) {
+        return false;
+      }
 
-  FindInstruction(message_.instruction_to_insert_before(), ir_context)
-      ->InsertBefore(MakeUnique<opt::Instruction>(
-          ir_context, static_cast<SpvOp>(message_.opcode()),
-          MaybeGetResultType(ir_context), message_.fresh_id(), in_operands));
+      if (const auto* vector = type->AsVector()) {
+        // Check that the operand has valid type.
+        if (!vector->element_type()->AsInteger() ||
+            vector->element_type()->AsInteger()->IsSigned() !=
+                (opcode == SpvOpConvertSToF)) {
+          return false;
+        }
 
-  ir_context->InvalidateAnalysesExceptFor(opt::IRContext::kAnalysisNone);
+        // Check that result type id exists in the module.
+        auto element_type_id = fuzzerutil::MaybeGetFloatType(
+            ir_context, vector->element_type()->AsInteger()->width());
+        return element_type_id != 0 &&
+               fuzzerutil::MaybeGetVectorType(ir_context, element_type_id,
+                                              vector->element_count()) != 0;
+      } else {
+        // Check that result type id exists in the module.
+        return type->AsInteger() &&
+               type->AsInteger()->IsSigned() == (opcode == SpvOpConvertSToF) &&
+               fuzzerutil::MaybeGetFloatType(ir_context,
+                                             type->AsInteger()->width()) != 0;
+      }
+    }
+    case SpvOpConvertFToU:
+    case SpvOpConvertFToS: {
+      if (message_.in_operand_id_size() != 1) {
+        return false;
+      }
 
-  transformation_context->GetFactManager()->AddFactIdEquation(
-      message_.fresh_id(), static_cast<SpvOp>(message_.opcode()), rhs_id,
-      ir_context);
-}
+      const auto* type = ir_context->get_type_mgr()->GetType(
+          fuzzerutil::GetTypeId(ir_context, message_.in_operand_id(0)));
+      if (!type) {
+        return false;
+      }
 
-protobufs::Transformation TransformationEquationInstruction::ToMessage() const {
-  protobufs::Transformation result;
-  *result.mutable_equation_instruction() = message_;
-  return result;
-}
+      if (const auto* vector = type->AsVector()) {
+        if (const auto* float_type = vector->element_type()->AsFloat()) {
+          // Check that result type id exists in the module.
+          auto element_type_id = fuzzerutil::MaybeGetIntegerType(
+              ir_context, float_type->width(), opcode == SpvOpConvertFToS);
```

Comment:
Again, I think it's OK for ConvertFToU to have a signed return type.  Seems weird, I know, but the point is that the instruction converts to an unsigned bit pattern - but that can be stored in a signed or unsigned id.

If this is true, it would be better to be more liberal here and do MaybeGetIntegerType with the expected sign, and then with the other sign - and only give up if neither exists.

Reply:
Fixed in a6ad6b3.